examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1159 @@
+      const std::vector<double> &JxW = scratch_data.get_JxW_values();
+
+      std::vector<double> rhs(n_q_points);
+      rhs_function.value_list(q_points, rhs);
+
+      for (unsigned int point = 0; point < n_q_points; ++point)
+        for (unsigned int i = 0; i < fe_v.dofs_per_cell; ++i)
+          {
+            for (unsigned int j = 0; j < fe_v.dofs_per_cell; ++j)
+              copy_data.cell_matrix(i, j) +=
+                // nu \nabla u \nabla v
+                viscosity * fe_v.shape_grad(i, point) *
+                fe_v.shape_grad(j, point) * JxW[point];
+
+            copy_data.cell_rhs(i) +=
+              rhs[point] * fe_v.shape_value(i, point) * JxW[point];
+          }
+    };
+
+    auto boundary_worker = [&](const Iterator &    cell,
+                               const unsigned int &face_no,
+                               ScratchData &       scratch_data,
+                               CopyData &          copy_data) {
+      const FEFaceValuesBase<dim> &fe_fv = scratch_data.reinit(cell, face_no);
+
+      const auto &       q_points      = scratch_data.get_quadrature_points();
+      const unsigned int n_q_points    = q_points.size();
+      const unsigned int dofs_per_cell = fe_fv.dofs_per_cell;
+
+      const std::vector<double> &        JxW = scratch_data.get_JxW_values();
+      const std::vector<Tensor<1, dim>> &normals =
+        scratch_data.get_normal_vectors();
+
+      std::vector<double> g(n_q_points);
+      boundary_function.value_list(q_points, g);
+
+
+      const double extent1 = cell->extent_in_direction(
+        GeometryInfo<dim>::unit_normal_direction[face_no]);
+      const double penalty = compute_penalty(extent1, extent1);
+
+      for (unsigned int point = 0; point < n_q_points; ++point)
+        {
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              copy_data.cell_matrix(i, j) +=
+                (
+                  // - nu (\nabla u . n) v
+                  -viscosity * (fe_fv.shape_grad(j, point) * normals[point]) *
+                    fe_fv.shape_value(i, point)
+
+                  // - nu u (\nabla v . n)  // NIPG: use +
+                  - viscosity * fe_fv.shape_value(j, point) *
+                      (fe_fv.shape_grad(i, point) * normals[point])
+
+                  // + nu * penalty u v
+                  + viscosity * penalty * fe_fv.shape_value(j, point) *
+                      fe_fv.shape_value(i, point)) *
+                JxW[point];
+
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            copy_data.cell_rhs(i) +=
+              (
+                // -nu g (\nabla v . n) // NIPG: use +
+                -viscosity * g[point] *
+                  (fe_fv.shape_grad(i, point) * normals[point])
+
+                // +nu penalty g v
+                +
+                viscosity * penalty * g[point] * fe_fv.shape_value(i, point)) *
+              JxW[point];
+        }
+    };
```

Comment:
Your implementation seems perfectly coherent with the equations you have above. This is the same exact implementation I had in my own DG Poisson equation :)! Really reassuring.